pkg/apis/pipeline/v1alpha1/resource_types.go

```diff
@@ -164,3 +164,39 @@ func ResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
 	}
 	return nil, fmt.Errorf("%s is an invalid or unimplemented PipelineResource", r.Spec.Type)
 }
+
+// AttributesFromType returns a list of available attributes that can be replaced
+// through templating.
+func AttributesFromType(prt PipelineResourceType) ([]string, error) {
+	r := &PipelineResource{}
+	r.Spec.Type = prt
+	// Todo (mk): We have run into this unfortunate event where technically we cannot determine
+	// what attributes an object has strictly from the PipelineResourceType alone.
+	// This is because of the subtypes that exists for the PipelineResourceTypeStorage type.
+	// At the moment, both of the subtypes happen to have equivalent keys in their
+	// Replacements(). So technically we can just pick one of the subtypes and still return the
+	// proper attributes. However, we may want to reconsider the approach on how we obtain or
+	// store these attributes.
+	if prt == PipelineResourceTypeStorage {
+		r.Spec.Params = []Param{
+
+			{
+				Name:  "type",
+				Value: string(PipelineResourceTypeGCS),
```

Comment:
`PipelineResourceTypeStorage` could be something else than `PipelineResourceTypeGCS` though

Reply:
Correct, I think this is the blocking issue at the moment. I left a comment above in the code which goes into more detail. 

>We have run into this unfortunate event where technically we cannot determine what attributes an object has strictly from the PipelineResourceType alone. This is because of the subtypes that exists for the PipelineResourceTypeStorage type. At the moment, both of the subtypes happen to have equivalent keys in their Replacements(). So technically we can just pick one of the subtypes and still return the proper attributes. However, we may want to reconsider the approach on how we obtain or store these attributes.

Still not exactly sure how to go about solving this problem, but I will think about it some more and update this PR with a better solution once I find the time.

Reply:
It seems like there isn't enough information in a `TaskResource` struct to correctly infer the type of a `PipelineResourceTypeStorage`. The exact storage type only gets determined when a `TaskRun` is created I think, right?  It doesn't seem to me like there's a way to validate storage when the Task is applied since its type parameter wont be declared until later.

Options I can see right now are:
1. Leave the code as it is :- make an assumption about the storage type. Future storage types will need to include these keys to validate or will need to change this validation code.
2. Augment the `TaskResource` with enough info to infer the correct storage type to validate against. This seems counter to the design of PipelineResourceTypeStorage - the dynamic binding of the storage's exact type is intentionally left until the TaskRun to allow different sources to feed through a pipeline, I think?
3. Agree on a set of templatable params that all storage types must implement.
4. Iterate over all known storage types and check whether the templated fields match at least one of those storage kinds.
5. Abandon validating storage params in a Task template.  Maybe do it in TaskRuns instead.

Have I misunderstood the implementation of `TaskResource` / `TaskResourceBinding`?  Are there other options available to us here?

Reply:
Your analysis seems correct to me and I was thinking about a few of those options as well. 

I agree that option 2 seems to counter the current design, though it's possible I have a misunderstanding.

I think the rest of the options you mentioned are solid approaches.

In my mind, option 3 sounds like an ideal approach and a convenient characteristic for all types of `PipelineResourceTypeStorage` to share. However, this might not be realistic depending on the types of `PipelineResourceTypeStorage` we plan to support in the future.

Option 4 seems like something that could be built now with not much effort and provide immediate benefit. There could be some cases in the future where different 
 types of `PipelineResourceTypeStorage` don't share the same templatable params, in which case we would only be able to catch certain errors at `TaskRun` execution time.

Option 5 seems reasonable and provides similar benefits as option 4. We will catch most kinds of templating errors earlier on but still have certain errors at `TaskRun` execution time.

---

After some thought, if we feel we could reach a consensus on option 3 then that seems like the most robust approach. 

If that's not something we think we can enforce, then option 4 or 5 sound great to me.

Reply:
Option 3 seems like a design that would be worth raising with the community - I doubt we could confidently come to a widely agreeable consensus in this github thread alone.  So the process for that would be to create a github issue describing the problem along with a suggested resolution and design doc if the problem is gnarly enough (this one is, I think).

It may also be that we can push this specific problem back until there is more movement on [Pipeline Resource Extensibility](https://github.com/tektoncd/pipeline/issues/238), which may end up setting the "validatable surface area" simply as a byproduct of the interfaces that it defines for extensions.

Having said all that it seems to me that we could move ahead with options 4 or 5 in this PR and work on option 3 out-of-band.  If it were me I'd probably skip validating storage types completely, leave a comment explaining why, and create a github issue to describe the problem and possible solutions in detail.

Reply:
Actually thinking about this some more, I reckon simply pushing the validation to the TaskRun is totally fine.  The whole point of the `PipelineResourceTypeStorage` is that it's a placeholder to be filled in later. The Task shouldn't need to really worry about validating it deeply.  At least from my POV.

Reply:
I'm now realizing that option 5, skipping storage params, is kind of tricky. It's not possible to easily express "ignore variables of this type" when doing the template validation.

Reply:
Ah, if the task shouldn't have to worry about validating it deeply, then it's possible that we do not really want this kind of patch and may just want to revisit this in a more broad way with a github issue and a design doc at some point.

Otherwise, it might be possible to go with a hybrid approach where we deeply validate all types of variables besides storage. Then for storage params, we just validate the top level variable (similar to how the validation works currently). But yeah, not sure if there is an easy way to express that as you said.

Been busy so I haven't had time to take a look, but I can take a look this weekend to see if i can come up with a nice way to solve it.